File: netLib/MyGraphApp/GraphLib/Graph.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using GraphLibrary.Representations;

namespace GraphLibrary
{
    public class Graph
    {
        private readonly IGraphRepresentation _representation;
        private List<int> _allDegreesCache;

        public int VertexCount => _representation.VertexCount;
        public int EdgeCount => _representation.EdgeCount;

        public Graph(int vertexCount, Func<int, IGraphRepresentation> representationFactory)
        {
            if (vertexCount <= 0)
                throw new ArgumentException("O número de vértices deve ser positivo.", nameof(vertexCount));
            
            _representation = representationFactory(vertexCount);
        }

        private void InvalidateDegreeCache() => _allDegreesCache = null;

        public void AddEdge(int u, int v)
        {
            if (u < 1 || u > VertexCount || v < 1 || v > VertexCount)
                throw new ArgumentOutOfRangeException($"Vértices devem estar no intervalo [1, {VertexCount}].");
            
            _representation.AddEdge(u, v);
            InvalidateDegreeCache();
        }

        public Dictionary<string, double> GetDegreeMetrics()
        {
            if (_allDegreesCache == null)
            {
                _allDegreesCache = new List<int>(VertexCount);
                for (int i = 1; i <= VertexCount; i++) _allDegreesCache.Add(_representation.GetNeighbors(i).Count());
            }
            var degrees = _allDegreesCache;

            if (!degrees.Any()) return new Dictionary<string, double> { { "min_degree", 0 }, { "max_degree", 0 }, { "avg_degree", 0 }, { "median_degree", 0 } };
            
            var sortedDegrees = degrees.OrderBy(d => d).ToList();
            double median = (sortedDegrees.Count % 2 == 1)
                ? sortedDegrees[sortedDegrees.Count / 2]
                : (sortedDegrees[sortedDegrees.Count / 2 - 1] + sortedDegrees[sortedDegrees.Count / 2]) / 2.0;

            return new Dictionary<string, double> { { "min_degree", degrees.Min() }, { "max_degree", degrees.Max() }, { "avg_degree", degrees.Average() }, { "median_degree", median } };
        }

        // --- OPTIMIZED SEARCH METHODS ---

        // Public version for convenience (allocates its own collections)
        public (Dictionary<int, int?> Parents, Dictionary<int, int> Levels) BreadthFirstSearch(int startVertex)
        {
            var parents = new Dictionary<int, int?>(VertexCount);
            var levels = new Dictionary<int, int>(VertexCount);
            BreadthFirstSearch(startVertex, parents, levels); // Call the high-performance version
            return (parents, levels);
        }

        // Internal version for high-performance loops (reuses collections)
        internal void BreadthFirstSearch(int startVertex, Dictionary<int, int?> parents, Dictionary<int, int> levels)
        {
            parents.Clear();
            levels.Clear();
            var queue = new Queue<int>();

            for (int i = 1; i <= VertexCount; i++) levels[i] = -1;

            levels[startVertex] = 0;
            parents[startVertex] = null;
            queue.Enqueue(startVertex);

            while (queue.Count > 0)
            {
                int u = queue.Dequeue();
                foreach (var v in _representation.GetNeighbors(u))
                {
                    if (levels[v] == -1)
                    {
                        levels[v] = levels[u] + 1;
                        parents[v] = u;
                        queue.Enqueue(v);
                    }
                }
            }
        }

        public (Dictionary<int, int?> Parents, Dictionary<int, int> Levels) DepthFirstSearch(int startVertex)
        {
            var parents = new Dictionary<int, int?>(VertexCount);
            var levels = new Dictionary<int, int>(VertexCount);
            DepthFirstSearch(startVertex, parents, levels);
            return (parents, levels);
        }

        internal void DepthFirstSearch(int startVertex, Dictionary<int, int?> parents, Dictionary<int, int> levels)
        {
            parents.Clear();
            levels.Clear();
            var visited = new HashSet<int>();
            var stack = new Stack<(int Vertex, int Level)>();

            for (int i = 1; i <= VertexCount; i++) levels[i] = -1;

            stack.Push((startVertex, 0));
            visited.Add(startVertex);
            parents[startVertex] = null;
            levels[startVertex] = 0;

            while (stack.Count > 0)
            {
                var (u, level) = stack.Pop();
                foreach (var v in _representation.GetNeighbors(u))
                {
                    if (visited.Add(v))
                    {
                        parents[v] = u;
                        levels[v] = level + 1;
                        stack.Push((v, level + 1));
                    }
                }
            }
        }

        public int GetDistance(int u, int v)
        {
            var (_, levels) = BreadthFirstSearch(u); // Uses the convenient public version
            return levels.GetValueOrDefault(v, -1);
        }

        public int GetDiameter()
        {
            if (VertexCount == 0) return 0;
            int maxDist = 0;
            // Reusing collections here as well for a significant boost
            var parents = new Dictionary<int, int?>(VertexCount);
            var levels = new Dictionary<int, int>(VertexCount);
            for (int i = 1; i <= VertexCount; i++)
            {
                BreadthFirstSearch(i, parents, levels);
                int currentMax = 0;
                foreach(var level in levels.Values) if(level > currentMax) currentMax = level;
                if (currentMax > maxDist) maxDist = currentMax;
            }
            return maxDist;
        }

        public int GetApproximateDiameter()
        {
            if (VertexCount == 0) return 0;
            var random = new Random();
            int s = random.Next(1, VertexCount + 1);
            var (_, levelsFromS) = BreadthFirstSearch(s);
            int u = levelsFromS.OrderByDescending(kvp => kvp.Value).First().Key;
            var (_, levelsFromU) = BreadthFirstSearch(u);
            return levelsFromU.Values.Max();
        }

        public List<List<int>> GetConnectedComponents()
        {
            var components = new List<List<int>>();
            var visited = new HashSet<int>();
            for (int i = 1; i <= VertexCount; i++)
            {
                if (visited.Add(i))
                {
                    var currentComponent = new List<int>();
                    var queue = new Queue<int>();
                    queue.Enqueue(i);
                    while (queue.Count > 0)
                    {
                        int u = queue.Dequeue();
                        currentComponent.Add(u);
                        foreach (var v in _representation.GetNeighbors(u)) if (visited.Add(v)) queue.Enqueue(v);
                    }
                    components.Add(currentComponent);
                }
            }
            return components.OrderByDescending(c => c.Count).ToList();
        }

        public static Graph FromFile(string filePath, Func<int, IGraphRepresentation> representationFactory)
        {
            var lines = File.ReadAllLines(filePath);
            if (!int.TryParse(lines.FirstOrDefault(), out int vertexCount))
                throw new InvalidDataException("A primeira linha deve conter o número de vértices.");

            var graph = new Graph(vertexCount, representationFactory);
            foreach (var line in lines.Skip(1))
            {
                var parts = line.Split(new[] { ' ', '\t' }, StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length == 2 && int.TryParse(parts[0], out int u) && int.TryParse(parts[1], out int v))
                {
                    graph.AddEdge(u, v);
                }
            }
            return graph;
        }
    }
}

File: lib/graph.py
import collections
import numpy as np
from representations import GraphRepresentation, AdjacencyList, AdjacencyMatrix
from typing import Type, Optional, Dict, List, Tuple

class Graph:
    """
    Classe principal para manipulação de grafos. Abstrai a representação
    subjacente e fornece uma interface de alto nível para o usuário.
    """
    def __init__(self, num_vertices: int, representation_cls: Type[GraphRepresentation] = AdjacencyList):
        if num_vertices <= 0:
            raise ValueError("O número de vértices deve ser um inteiro positivo.")
        
        self._num_vertices = num_vertices
        self._representation = representation_cls(num_vertices)
        self._all_degrees_cache = None # Cache para os graus

    def _invalidate_degree_cache(self):
        """Invalida o cache de graus sempre que uma aresta é adicionada."""
        self._all_degrees_cache = None

    @property
    def vertex_count(self) -> int:
        return self._num_vertices

    @property
    def edge_count(self) -> int:
        return self._representation.edge_count

    def add_edge(self, u: int, v: int):
        if not (GraphUtils.is_vertex_within_bounds(u, self._num_vertices) and GraphUtils.is_vertex_within_bounds(v, self._num_vertices)):
            raise ValueError(f"Vértices {u} e {v} estão fora do intervalo [1, {self._num_vertices}]")
        
        self._representation.add_edge(u, v)
        self._invalidate_degree_cache()

    def get_neighbors(self, v: int) -> List[int]:
        """Retorna uma lista de vizinhos do vértice v."""
        if not GraphUtils.is_vertex_within_bounds(v, self._num_vertices):
            raise ValueError(f"Vértice {v} está fora do intervalo [1, {self._num_vertices}]")
        
        return self._representation.get_neighbors(v)

    # --- Métricas de Grau ---

    def _calculate_all_degrees(self) -> List[int]:
        """Calcula e cacheia os graus de todos os vértices."""
        if self._all_degrees_cache is None:
            self._all_degrees_cache = [len(self.get_neighbors(i)) for i in range(1, self.vertex_count + 1)]
        return self._all_degrees_cache

    def get_degree_metrics(self) -> Dict[str, float]:
        """
        Retorna um dicionário com o grau mínimo, máximo, médio e a mediana de grau. [cite: 107]
        """
        degrees = self._calculate_all_degrees()
        sorted_degrees = sorted(degrees)
        
        n = len(sorted_degrees)
        median = 0
        if n > 0:
            if n % 2 == 1:
                median = sorted_degrees[n // 2]
            else:
                mid1 = sorted_degrees[n // 2 - 1]
                mid2 = sorted_degrees[n // 2]
                median = (mid1 + mid2) / 2

        return {
            "min_degree": min(degrees) if degrees else 0,
            "max_degree": max(degrees) if degrees else 0,
            "avg_degree": np.mean(degrees) if degrees else 0,
            "median_degree": median
        }
        
    # --- Algoritmos de Busca ---

    def breadth_first_search(self, start_vertex: int) -> Tuple[Dict[int, Optional[int]], Dict[int, int]]:
        """
        Executa a Busca em Largura (BFS) a partir de um vértice inicial.

        Args:
            start_vertex: O vértice para iniciar a busca.

        Returns:
            Uma tupla contendo dois dicionários:
            - parents: Mapeia cada vértice ao seu pai na árvore de busca.
            - levels: Mapeia cada vértice ao seu nível (distância da raiz).
        """
        if not (GraphUtils.is_vertex_within_bounds(start_vertex, self.vertex_count)):
            raise ValueError(f"Vértice inicial {start_vertex} está fora do intervalo.")

        parents = {v: None for v in range(1, self.vertex_count + 1)}
        levels = {v: -1 for v in range(1, self.vertex_count + 1)}
        
        queue = collections.deque([start_vertex])
        levels[start_vertex] = 0
        
        while queue:
            u = queue.popleft()
            for v in self.get_neighbors(u):
                if levels[v] == -1: # Se o vértice não foi visitado
                    levels[v] = levels[u] + 1
                    parents[v] = u
                    queue.append(v)
        
        return parents, levels

    def depth_first_search(self, start_vertex: int) -> Tuple[Dict[int, Optional[int]], Dict[int, int]]:
        """
        Executa a Busca em Profundidade (DFS) a partir de um vértice inicial. [cite: 112]

        Args:
            start_vertex: O vértice para iniciar a busca.

        Returns:
            Uma tupla contendo dois dicionários:
            - parents: Mapeia cada vértice ao seu pai na árvore de busca.
            - levels: Mapeia cada vértice à sua profundidade na árvore de busca.
        """
        if not (1 <= start_vertex <= self.vertex_count):
            raise ValueError(f"Vértice inicial {start_vertex} está fora do intervalo.")
            
        parents = {v: None for v in range(1, self.vertex_count + 1)}
        levels = {v: -1 for v in range(1, self.vertex_count + 1)}
        stack = [(start_vertex, 0)] # Armazena (vértice, nível)
        
        visited = {start_vertex}
        parents[start_vertex] = None
        levels[start_vertex] = 0
        
        while stack:
            u, level = stack.pop()
            
            for v in sorted(self.get_neighbors(u), reverse=True): # Ordem consistente
                if v not in visited:
                    visited.add(v)
                    parents[v] = u
                    levels[v] = level + 1
                    stack.append((v, level + 1))
                    
        return parents, levels

    # --- Distância e Diâmetro ---

    def get_distance(self, u: int, v: int) -> int:
        """
        Calcula a distância (caminho mínimo) entre dois vértices. 
        Retorna -1 se não houver caminho.
        """
        _, levels = self.breadth_first_search(u)
        return levels[v]

    def get_diameter(self) -> int:
        """
        Calcula o diâmetro exato do grafo. [cite: 115, 116]
        Pode ser computacionalmente caro para grafos grandes.
        """
        max_dist = 0
        for i in range(1, self.vertex_count + 1):
            _, levels = self.breadth_first_search(i)
            current_max = max(levels.values())
            if current_max > max_dist:
                max_dist = current_max
        return max_dist

    # --- Componentes Conexas ---
    
    def get_connected_components(self) -> List[Dict]:
        """
        Encontra todas as componentes conexas do grafo. [cite: 117]
        Retorna uma lista de dicionários, cada um representando uma componente,
        ordenada por tamanho em ordem decrescente. 
        """
        visited = set()
        components = []
        for v in range(1, self.vertex_count + 1):
            if v not in visited:
                component_nodes = set()
                queue = collections.deque([v])
                visited.add(v)
                component_nodes.add(v)
                
                while queue:
                    u = queue.popleft()
                    for neighbor in self.get_neighbors(u):
                        if neighbor not in visited:
                            visited.add(neighbor)
                            component_nodes.add(neighbor)
                            queue.append(neighbor)
                
                components.append({
                    "size": len(component_nodes),
                    "vertices": sorted(list(component_nodes))
                })
        
        # Ordena as componentes por tamanho, em ordem decrescente
        return sorted(components, key=lambda x: x['size'], reverse=True)

    # --- Método de Fábrica ---

    @staticmethod
    def from_file(file_path: str, representation_cls: Type[GraphRepresentation] = AdjacencyList):
        try:
            with open(file_path, 'r') as f:
                num_vertices = int(f.readline().strip())
                graph = Graph(num_vertices, representation_cls)

                for line in f:
                    parts = line.strip().split()
                    if len(parts) == 2:
                        u, v = int(parts[0]), int(parts[1])
                        graph.add_edge(u, v)
                return graph
        except FileNotFoundError:
            raise FileNotFoundError(f"Arquivo não encontrado em: {file_path}")
        except ValueError:
            raise ValueError("Formato de arquivo inválido. Verifique o conteúdo.")
        

class GraphUtils:

    @staticmethod
    def is_vertex_within_bounds(v: int, num_vertices: int) -> bool:
        """Verifica se um vértice está dentro dos limites válidos."""
        return 1 <= v <= num_vertices